### ISSUE TITLE
ResultForceHTTPS: Check if https:// works before replacing links to https://

### ISSUE THREAD

**Issue by DarkDeviL** [ShareX/ShareX#2559]
As more and more sites are moving towards https://, I agree that having the option to provide https:// links (**if they work!**) is a great option.

However, I don't think that @mileswd (#689), @Slind14 (#716), @AustinPaquette (#793), @Phoenix616 (#904), @redpois0n (#989), @Botlover (#1027), @daofresh (#1274), @SpyTec (#1641), @darktohka (#1781), @JetCrusherTorpedo (#1852, #1940) or @MorkHub (#2113) wanted to break functionality with their suggestion(s) and/or pull request(s).

I would therefore suggest that if the option is set to true, and the image URL received is:
```
http://example.net/img/test.png
```

Then ShareX should try fetching the HTTPS version, e.g.: 
```
https://example.net/img/test.png
```

Following would then be applied:

1. If download succeeds (e.g. no SSL warnings, etc.), then:
2. Check MD5/SHA1/[...you name it] hash, file size, etc. to see if it matches the original uploaded contents.

If any of it fails, or that the SSL certificate is invalid (expired, invalid domain, not trusted CA, etc.), then the original, unedited and insecure HTTP link would be returned.

Description phrase of the Task setting could then be rewritten into something like:

```
If result URL starts with "http://", then ShareX will check whether the "https://" address works, and if it does, the URL will be replaced with it's "https://" version.
If ShareX still returns an "http://" URL with this setting set to True, you need to contact the upload destination, and convince them to enable HTTPS for their domain(s).
```

The point is it is more sane to check whether https:// actually works, before providing the https:// URL, even with the setting set to True.

In short words, replacing the function's functionality with "Force HTTPS (**if available**)".

**Comment by DarkDeviL** [comment]
:(

**Comment by DarkDeviL** [comment]
@stuntguy3000: 

[In this one](https://github.com/ShareX/ShareX/pull/689#issuecomment-102710286), @Jaex wouldn't introduce something that would break the "official URL format".

´´´
I can't risk using non official URL format.
´´´

If a site is not accessible over HTTPS://, @Jaex's original statement doesn't hold, and this setting is supporting "non official URL formats", as @Jaex wouldn't do. @Jaex is referencing that issue and the same phrase over and over, in most of the references.

If you really feel like following standards and/or the official ways provided by the third parties you are working with, in your projects, are beyond the scope of your projects, then I feel very sorry for your projects, @stuntguy3000!

Since there are 9 thumbs up (10, if we are counting my "issue" as one too), but you still had to comment "What's the issue?", really makes it seem like you should try reading reading things, before commenting, in the future....

**Comment by stuntguy3000** [comment]
It's 2018, if a site isn't using HTTPS, or at-least doesn't support it, **then stop using it**

I know that's what you want to hear, but the FOSS market has to discourage HTTP only where possible, and this is one instance of ShareX and Jaex taking the progressive stance.

**Comment by DarkDeviL** [comment]
If I had a chance of ensuring that 100% of the web would use HTTPS only, with proper configurations and certificates, and so on, - please believe me, I would take that chance!

The concept of ResultForceHTTPS is perfectly okay to me, but it wouldn't take much to actually try the URL, if you receive
```
http://example.net/img/test.png
```
then just make a simple GET request to:
```
https://example.net/
```
If it works without any HTTPS errors, etc., and there is a proper HTTP response from the server via HTTPS, switch it to HTTPS, otherwise don't.

If you really want to take go the "progressive" way, to use your word, you could the warn the user with a an alert box, every time HTTPS isn't possible, mentioning that they should consult their upload provider, and request them to add HTTPS to their servers, and mention that if their provider adds HTTPS, the alert box would automatically go away. That way would help the situation much more. Example.net's support desk would receive queries from their users, and if the amount of queries are taking up a decent amount of their time, they will most likely add it, in order to reduce the time they need to spend on their support desk.

People usually hate warnings, errors and alerts, and that one would be much more "progressive".

I can't decide whether ShareX should just be "cool", or if the day where it would become "super cool" is near. I'm just trying to help it to become "super cool", that's all.

**Comment by sylveon** [comment]
Doing said request also incurs bigger upload times.

**Comment by DarkDeviL** [comment]
@sylveon Bigger upload times? As the stuff is already uploaded, when the application would check if the HTTPS version of the URL works, it obviously won't be doing any changes to the upload time itself.

It will add a tiny bit of latency, which is quite often around 0.2 - 0.5 seconds for most destinations with the HTTP check depending on the server, though cross-continent requests may be a little bit larger though, plus of course the kind of latency ShareX itself would be adding on top of that...

Doing a simple HTTP call to "https://i.imgur.com/", which is contained in several of the references above, seem to be a latency of 0.157s in Ukraine (maximum in Europe), and 0.362s in Vietnam according to https://check-host.net - from my end, the total processing time is 0.159s, with some old 2001 hardware...
